### ISSUE TITLE
Linux / Jackd: Allow external connection management

### ISSUE THREAD

**Issue by mark-orion** [sonosaurus/sonobus#152]
I am using SonoBus with jackd on Ubuntu Studio and I notice that SonoBus only works with jackd if the software itself creates the jackd connections. It is not possible for another application like QjackCtl or Ardour to manage the SonoBus jackd ports. It would be good if SonoBus could just present jackd ports without the autoconnection feature. My suggestion would be to do this when << none >> is selected as jackd output or input device.

**Comment by essej** [comment]
I know, the way JUCE presents JACK in the standalone is really not suitable for the best use of JACK. It would take some redesign work to present it more as a plugin, where connections are made to it instead of focusing on the hardware I/O.

**Comment by strk** [comment]
I'm not sure this is the appropriate ticket to request this feature or if I should file a new issue but would a redesign also allow us to define an arbitrary number of input ports ? I'm asking because at the moment when selecting JACK as "Audio device type" I'm limited to use ONLY the number of actual hardware inputs I have, while I'd like to use more inputs to sonobus for virtual channels

**Comment by essej** [comment]
> I'm not sure this is the appropriate ticket to request this feature or if I should file a new issue but would a redesign also allow us to define an arbitrary number of input ports ? I'm asking because at the moment when selecting JACK as "Audio device type" I'm limited to use ONLY the number of actual hardware inputs I have, while I'd like to use more inputs to sonobus for virtual channels

Yes, this is why the current JUCE implementation is really not good, and yes the same redesign would address that. You could try to suggest this over on the JUCE forum... but I doubt it will make it to the top of their list very soon at all. It will probably take someone like me taking the time to just implement it... and *maybe* they would accept it...

**Comment by strk** [comment]
would NSM protocol be part of the redesign ?

**Comment by strk** [comment]
Someone pointed me to https://github.com/guysherman/jack-passthrough as a workaround until JUCE is fixed to deal properly with Jack

**Comment by strk** [comment]
Same problem with JUCE from another application: https://github.com/BespokeSynth/BespokeSynth/issues/273

**Comment by strk** [comment]
Upstream ticket: https://github.com/juce-framework/JUCE/issues/333

**Comment by riban-bw** [comment]
I am not sure why this is a JUCE issue. Many JUCE based applications and plugins define the quantity of audio inputs and outputs and allow the plugin host or other jack graph manager to manage interconnects. The difference here is that the standalone version of sonobus allows selection of which devices to connect to (hardware ports, other jack clients, etc.) and the LV2 plugin just creates a load of ports. I don't understand why sonobus needs all these ports. It seems to create a similar quantity of ports as the selected device has, e.g. selecting system will create 22 ports on my system because I have a soundcard with 16 audio and 6 MIDI ports. (It should be ignoreing MIDI ports - but that is another bug report!)

Regarding the subject of this feature request, both the standalone and LV2 plugin allows management of the jack graph by external application, e.g. I can route using patachage and sonobus recognises that the routes have changed.

I do not understand why sonobus creates so many audio ports but suspect it is to allow individual inputs to be fed over a multichannel AoIP link. I think the assumption that this should be equal to the quantity of hardware ports is flawed. The user should be able to define the quantity of input and output ports to suit their needs. I may have 16 physical input ports but I may wish to only use 2 sonobus ports or I may want to feed all 16 plus another 4 from another jack source.

I searched for issues relating to the quantity of ports which this ticket isn't really related to but the subject was raised here, hence my comments.
